File: app/services/metadata_service.py
import logging
from uuid import UUID
from typing import Optional

from sqlalchemy import text
from sqlalchemy.orm import Session

from app.database import create_target_engine
from app.models import DatabaseMetadata, TableMetadata, ColumnMetadata

logger = logging.getLogger("app.services.metadata")


class MetadataService:
    """Service for extracting and managing database metadata."""

    def __init__(self, db: Session):
        self.db = db

    def extract_and_store_metadata(
        self,
        host: str,
        port: int,
        database: str,
        username: str,
        password: str,
    ) -> DatabaseMetadata:
        """
        Connect to a target PostgreSQL database, extract table/column metadata,
        and store it in the system database.
        """
        # Security: mask sensitive info in logs
        masked_user = f"{username[:2]}***" if len(username) > 2 else "***"
        logger.info(f"Connecting to target database: {database}@{host}:{port} (user: {masked_user})")

        # Create connection to target database
        target_engine = create_target_engine(host, port, database, username, password)

        # Test connection
        try:
            with target_engine.connect() as conn:
                conn.execute(text("SELECT 1"))
            logger.info(f"Successfully connected to target database: {database}")
        except Exception as e:
            logger.error(f"Failed to connect to target database {database}@{host}:{port}: {type(e).__name__}")
            raise ConnectionError(
                f"Cannot connect to database '{database}' at {host}:{port}. "
                f"Please verify connection details. Error: {type(e).__name__}"
            )

        # Create database metadata record
        db_metadata = DatabaseMetadata(
            database_name=database,
            host=host,
            port=port,
            username=username,
            password=password,
        )
        self.db.add(db_metadata)
        self.db.flush()
        logger.info(f"Created metadata record: {db_metadata.id}")

        # Extract table and column information from information_schema
        try:
            with target_engine.connect() as conn:
                # Get all tables in public schema
                tables_result = conn.execute(
                    text("""
                        SELECT table_name 
                        FROM information_schema.tables 
                        WHERE table_schema = 'public' 
                          AND table_type = 'BASE TABLE'
                        ORDER BY table_name
                    """)
                )
                tables = tables_result.fetchall()
                logger.info(f"Found {len(tables)} tables in database '{database}'")

                for table_row in tables:
                    table_name = table_row[0]

                    # Create table metadata
                    table_metadata = TableMetadata(
                        database_id=db_metadata.id,
                        table_name=table_name,
                        schema_name="public",
                    )
                    self.db.add(table_metadata)
                    self.db.flush()

                    # Get column information for this table
                    columns_result = conn.execute(
                        text("""
                            SELECT 
                                column_name,
                                data_type,
                                is_nullable,
                                column_default,
                                ordinal_position
                            FROM information_schema.columns 
                            WHERE table_schema = 'public' 
                              AND table_name = :table_name
                            ORDER BY ordinal_position
                        """),
                        {"table_name": table_name},
                    )
                    columns = columns_result.fetchall()
                    logger.debug(f"  Table '{table_name}': {len(columns)} columns")

                    for col_row in columns:
                        column_metadata = ColumnMetadata(
                            table_id=table_metadata.id,
                            column_name=col_row[0],
                            data_type=col_row[1],
                            is_nullable=col_row[2],
                            column_default=col_row[3],
                            ordinal_position=col_row[4],
                        )
                        self.db.add(column_metadata)

        except Exception as e:
            logger.error(f"Failed to extract metadata from database '{database}': {e}")
            self.db.rollback()
            raise RuntimeError(
                f"Failed to extract metadata from database '{database}': {str(e)}"
            )

        self.db.commit()
        self.db.refresh(db_metadata)

        # Dispose the target engine connection
        target_engine.dispose()

        total_columns = sum(len(t.columns) for t in db_metadata.tables)
        logger.info(
            f"Metadata extraction complete: {db_metadata.table_count} tables, "
            f"{total_columns} columns stored with ID {db_metadata.id}"
        )

        return db_metadata

    def list_all_metadata(self) -> list[DatabaseMetadata]:
        """List all stored metadata records."""
        records = self.db.query(DatabaseMetadata).order_by(DatabaseMetadata.created_at.desc()).all()
        logger.info(f"Listed {len(records)} metadata records")
        return records

    def get_metadata_by_id(self, metadata_id: UUID) -> Optional[DatabaseMetadata]:
        """Get a specific metadata record by its ID."""
        record = self.db.query(DatabaseMetadata).filter(DatabaseMetadata.id == metadata_id).first()
        if record:
            logger.info(f"Found metadata record: {metadata_id}")
        else:
            logger.warning(f"Metadata record not found: {metadata_id}")
        return record

    def delete_metadata(self, metadata_id: UUID) -> bool:
        """Delete a metadata record and all associated data (cascading)."""
        record = self.db.query(DatabaseMetadata).filter(DatabaseMetadata.id == metadata_id).first()
        if not record:
            logger.warning(f"Cannot delete: metadata {metadata_id} not found")
            return False

        self.db.delete(record)
        self.db.commit()
        logger.info(f"Deleted metadata record: {metadata_id} (cascade: tables + columns)")
        return True

File: java-app/src/main/java/com/kafein/discovery/service/MetadataService.java
package com.kafein.discovery.service;

import com.kafein.discovery.dto.response.MetadataDetailResponse;
import com.kafein.discovery.dto.response.MetadataListResponse;
import com.kafein.discovery.entity.ColumnMetadata;
import com.kafein.discovery.entity.DatabaseMetadata;
import com.kafein.discovery.entity.TableMetadata;
import com.kafein.discovery.repository.DatabaseMetadataRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class MetadataService {

    private final DatabaseMetadataRepository repository;

    @Transactional
    public DatabaseMetadata extractAndStoreMetadata(
            String host, int port, String database, String username, String password) {

        String maskedUser = username.length() > 2 ? username.substring(0, 2) + "***" : "***";
        log.info("Connecting to target database: {}@{}:{} (user: {})", database, host, port, maskedUser);

        String jdbcUrl = String.format("jdbc:postgresql://%s:%d/%s", host, port, database);

        // Test connection
        try (Connection conn = DriverManager.getConnection(jdbcUrl, username, password)) {
            log.info("Successfully connected to target database: {}", database);
        } catch (SQLException e) {
            log.error("Failed to connect to target database {}@{}:{}: {}", database, host, port, e.getClass().getSimpleName());
            throw new RuntimeException(
                "Cannot connect to database '" + database + "' at " + host + ":" + port +
                ". Please verify connection details. Error: " + e.getClass().getSimpleName()
            );
        }

        // Create metadata record
        DatabaseMetadata dbMetadata = DatabaseMetadata.builder()
            .databaseName(database)
            .host(host)
            .port(port)
            .username(username)
            .password(password)
            .build();

        // Extract tables and columns using information_schema
        try (Connection conn = DriverManager.getConnection(jdbcUrl, username, password)) {
            // Get all tables in public schema
            String tableSql = """
                SELECT table_name FROM information_schema.tables
                WHERE table_schema = 'public' AND table_type = 'BASE TABLE'
                ORDER BY table_name
                """;

            try (PreparedStatement stmt = conn.prepareStatement(tableSql);
                 ResultSet rs = stmt.executeQuery()) {

                while (rs.next()) {
                    String tableName = rs.getString("table_name");

                    TableMetadata tableMetadata = TableMetadata.builder()
                        .database(dbMetadata)
                        .tableName(tableName)
                        .schemaName("public")
                        .build();

                    // Get columns for this table
                    String colSql = """
                        SELECT column_name, data_type, is_nullable, column_default, ordinal_position
                        FROM information_schema.columns
                        WHERE table_schema = 'public' AND table_name = ?
                        ORDER BY ordinal_position
                        """;

                    try (PreparedStatement colStmt = conn.prepareStatement(colSql)) {
                        colStmt.setString(1, tableName);
                        try (ResultSet colRs = colStmt.executeQuery()) {
                            List<ColumnMetadata> columns = new ArrayList<>();
                            while (colRs.next()) {
                                ColumnMetadata col = ColumnMetadata.builder()
                                    .table(tableMetadata)
                                    .columnName(colRs.getString("column_name"))
                                    .dataType(colRs.getString("data_type"))
                                    .isNullable(colRs.getString("is_nullable"))
                                    .columnDefault(colRs.getString("column_default"))
                                    .ordinalPosition(colRs.getInt("ordinal_position"))
                                    .build();
                                columns.add(col);
                            }
                            tableMetadata.setColumns(columns);
                        }
                    }

                    dbMetadata.getTables().add(tableMetadata);
                }
            }

            log.info("Found {} tables in database '{}'", dbMetadata.getTableCount(), database);

        } catch (SQLException e) {
            log.error("Failed to extract metadata from database '{}': {}", database, e.getMessage());
            throw new RuntimeException("Failed to extract metadata from database '" + database + "': " + e.getMessage());
        }

        DatabaseMetadata saved = repository.save(dbMetadata);
        int totalColumns = saved.getTables().stream().mapToInt(t -> t.getColumns().size()).sum();
        log.info("Metadata extraction complete: {} tables, {} columns stored with ID {}",
            saved.getTableCount(), totalColumns, saved.getId());

        return saved;
    }

    @Transactional(readOnly = true)
    public List<MetadataListResponse> listAllMetadata() {
        return repository.findAllByOrderByCreatedAtDesc().stream()
            .map(m -> new MetadataListResponse(
                m.getId(),
                m.getDatabaseName(),
                m.getCreatedAt(),
                m.getTableCount()
            ))
            .toList();
    }

    @Transactional(readOnly = true)
    public Optional<MetadataDetailResponse> getMetadataById(UUID id) {
        return repository.findById(id).map(this::toDetailResponse);
    }

    @Transactional
    public boolean deleteMetadata(UUID id) {
        if (repository.existsById(id)) {
            repository.deleteById(id);
            log.info("Deleted metadata record: {} (cascade: tables + columns)", id);
            return true;
        }
        log.warn("Cannot delete: metadata {} not found", id);
        return false;
    }

    private MetadataDetailResponse toDetailResponse(DatabaseMetadata m) {
        List<MetadataDetailResponse.TableResponse> tables = m.getTables().stream()
            .map(t -> new MetadataDetailResponse.TableResponse(
                t.getId(),
                t.getTableName(),
                t.getSchemaName(),
                t.getColumns().stream()
                    .map(c -> new MetadataDetailResponse.ColumnResponse(
                        c.getId(),
                        c.getColumnName(),
                        c.getDataType(),
                        c.getIsNullable(),
                        c.getColumnDefault(),
                        c.getOrdinalPosition()
                    ))
                    .toList()
            ))
            .toList();

        return new MetadataDetailResponse(
            m.getId(),
            m.getDatabaseName(),
            m.getTableCount(),
            tables
        );
    }
}
